### DISCUSSION TITLE
使用Nonebot2a16推出的消息模板`Message.template`来进行消息格式化!

### DISCUSSION THREAD

**Discussion by mnixry** [nonebot/discussions#41] [经验]
之前, 我在讨论版发起了一份帖子: https://github.com/nonebot/discussions/discussions/27, 是我自己实现的一款消息模板格式化方案, 它用起来就像这样~

```python
from nonebot.adapters.cqhttp import Message

Message.template('{}是我老婆').format('@minxry')
```

有人可能会问了, 这和我用`str.format`或者`f-string`有什么区别捏?

来, 让我给你看点好康的:

```python
await bot.send(event, '{nickname}是我老婆!'.format(nickname=event.sender.nickname))
# 或者用f-string
await bot.send(event, f'{event.sender.nickname}是我老婆!')
```

非常不巧地是, 有一个触发该指令的群友昵称叫`[CQ:at,qq=all]`
于是Bot就会发出:`@全体成员 是我老婆!`
这样可不行, Bot就被这位群友变成了共用RBQ~

咳咳, 正经来讲, 这一行为被习惯性地称作CQ码注入 (拓展阅读: https://github.com/nonebot/nonebot2/issues/212#issuecomment-778007483)
当你使用了基于`OneBot v11`的实现, 就有可能发生这种问题

那么我们的`MessageTemplate`表现如何?

```python
await bot.send(event, Message.template('{nickname}是我老婆!').format(nickname='[CQ:at,qq=all]'))
```

现在, 这位~~居心叵测~~的群员就没有办法让Bot~~恶堕~~成为RBQ辣!

而且, 它的功能不止如此!
它支持用输入的`Message`对象作为字符串模板!

```python
template = Message.template(MessageSegment.text('{nickname}是我老婆!') + MessageSegment.image('file:///C:/脸红.jpg'))
await bot.send(event, template.format(nickname=event.sender.nickname))
```

那么, 它就会发送一张带有`C:\脸红.jpg`图片的`xxx是我老婆`消息了!

而且, 还有而且! 它事实上继承了`str.format`的实现, 所以它支持format的控制符!
这意味着你可以用类似于`{score:.3%}`这样的来输出`100.000%`! 具体[可以参考Python官方文档](https://docs.python.org/zh-cn/3/library/string.html#formatspec)

而且, 居然还有而且! 它已经被内置在NoneBot的多个功能中! 比如

`matcher.got`，`matcher.pause`，`matcher.reject`中的`prompt`；`matcher.send`，`matcher.finish`中的`message`就支持使用`MessageTemplate`以及`state`字典中的数据进行format

```python
@matcher.handle()
async def _(bot: Bot, event: Event, state: T_State) -> None:
    state["at"] = MessageSegment.at(event.get_user_id())
    state["user"] = event.get_user_id()
    ft = Message.template("{at} Hello! {user}")
    await template.send(ft)
```

而且, 甚至还有而且! 我们计划在下个版本中加入类似`{link:image}`这样的支持! 直接把链接转化为图片! 听起来很棒是不是!

这里面的代码全是手打的! 颅内代码检查! 各位大佬不要打我qwq
就这样~

> 这是参考资料:
> - https://github.com/nonebot/nonebot2/pull/510
> - https://github.com/nonebot/nonebot2/pull/545

~~这次尝试了一下比较可爱(?)的写作风格, 希望大家不要`D区`一下被恶心吐出来!~~

**Comment by kexue-z** [comment]
诶~看不懂!

**Reply by mnixry** [reply]
呜, `'xxx{}'.format('')`这种总用过吧, 就差不多的样子~
但是更安全而且更好用了!

**Comment by j1g5awi** [comment]
网络美少女差不多得了（

**Reply by mnixry** [reply]
@mnixry 点了个踩

**Comment by AkiraXie** [comment]
所以大佬能不能整点使用的example出来，萌新不太明白场景

**Reply by mnixry** [reply]
看参考资料, 里面都有能用的demo

**Comment by mnixry** [comment]
> 我们计划在下个版本中加入类似{link:image}这样的支持! 直接把链接转化为图片!

该功能已添加到NoneBot2: https://github.com/nonebot/nonebot2/pull/555, 将于下个版本发布

~~发病完了感觉自己浓度好高, 好羞耻~~

**Comment by felinae98** [comment]
个人感觉这个功能上的太急了，结果就是插件如果适配了a16就不能在a15上工作

**Reply by felinae98** [reply]
除非对两个版本都做处理

**Reply by yanyongyu** [reply]
这个由插件发布者自行决定，你可以抛弃a16以下版本支持
